### ISSUE TITLE
Segfault when functionalization tensor wrapping subclass is an output saved for backward

### ISSUE THREAD

**Issue by soulitzer** [pytorch/pytorch#81365]
When the output of a function is a functionalization tensor and saved for backward, we call `.tensor_data()` which calls `shallow_copy_and_detach` underneath.

Repro:
```python
import torch
from torch.testing._internal.logging_tensor import LoggingTensor

inp = torch._to_functional_tensor(LoggingTensor(torch.randn(2, 2))).requires_grad_(True)
inp.exp()
```
Changing `exp` to `sigmoid` which saves output for bw also segfaults, but changing `sin` which saves input for bw is OK

The issue is that `pyobj_interpreter_` is `nullptr`, but still has Python key in https://github.com/pytorch/pytorch/blob/c2d395cf8e2d36bc012a13cdca6ef5bbcb14a0f1/aten/src/ATen/FunctionalTensorWrapper.cpp#L270-L292

Adding a check for `pyobj_interpreter_` seems to prevent segfault, but not sure if that is the right fix

cc @ezyang @gchanan @zou3519 @bdhirsh
